Design.md

```diff
@@ -0,0 +1,149 @@
+
+## CLI
+
+The CLI can be used to operate the job runner. All CLI commands have the job server URL as the first argument and the actual command after that.
+
+```
+cli https://job-service.example.org:8000 subcommand arguments
+```
+
+It has the following subcommands:
+
+* `start <executable> [args]...` -- Starts a new job by spawning a process, prints the job id to stdout.
+* `stop jobid` -- Stops job with given id.
+* `status jobid` -- Prints job status to stdout.
+* `output jobid` -- Streams job stdout and stderr to respective output streams. Starts from the beginning of the job.
+
+### Example usage
+
+#### List directory
+
+```
+$ cli https://localhost:8000 start ls /usr
+bba87bff-3719-4f76-98d9-c8e86f03f7aa
+$ cli https://localhost:8000 output bba87bff-3719-4f76-98d9-c8e86f03f7aa
+bin
+games
+include
+lib
+lib32
+libexec
+local
+sbin
+share
+src
+$ cli https://localhost:8000 status bba87bff-3719-4f76-98d9-c8e86f03f7aa
+Completed 0
+```
+
+#### Stopping a process
+
+```
+$ cli https://localhost:8000 start yes
+a354142a-c59f-44dd-ac53-c0110943df2b
+$ cli https://localhost:8000 output a354142a-c59f-44dd-ac53-c0110943df2b
+y
+y
+y
+y
+y
+...     # Omitted
+y
+y
+^C      # Keyboard interrupt
+$ cli https://localhost:8000 stop a354142a-c59f-44dd-ac53-c0110943df2b
+Signal 9
+$ cli https://localhost:8000 status a354142a-c59f-44dd-ac53-c0110943df2b
+Signal 9
+```
+
+
+## Scalability and high availability
+
+The system is reasonably perfomant, it should be able to handle quite a lot of processes on a single machine. However, there is no way to run this on multiple machines without some coordinator in front. The limits the horizontal scalability to a single machine. Of course it's possible for the client to handle multiple machines, but this is not built into the system. Ideally there would be a way to move processes from machine to another, but whether that's viable depends on many factors. This requires some way to synchronize shared state, usually using a database
+
+As for high availablity, similar concerns apply. Again, coordination with multiple controller nodes might be the best way to solve this. Many things depend on what properties are required. For instance, usually the platform itself cannot handle all high-availability concerns, as those must be solved by the application. However, things like "there should always be at least one copy of this software running" can usually be solved by duplication.
+
+### Other tradeoffs and simplifications
+
+Full output history of all jobs is stored in memory, and is only removed on server restart. In a real system, the output would usually be streamed to a log database, or just into a file, to reduce memory pressure. After process termination logs should be either removed or moved to an archive (e.g. Amazon S3).
```

Comment:
Wouldn't similar pressure concerns exist for a file on disk as it does on amount of RAM?

Reply:
They do, but you can usually have terabytes of disk fairly cheaply, while RAM is usually limited to (tens of) gigabytes. The correct approach depends on the schenario, as always.